### PR TITLE
Reland "[clang][modules] Print library module manifest path."

### PR DESCRIPTION
**Pull request by mordante**
This implements a way for the compiler to find the modules.json associated with the C++23 Standard library modules.

This is based on a discussion in SG15. At the moment no Standard library installs this manifest. #75741 adds this feature in libc++.

This reverts commit 82f424f766be00b037a706a835d0a0663a2680f1.

Disables the tests on non-X86 platforms as suggested.

### CODE REVIEW

File: clang/test/Driver/modules-print-library-module-manifest-path.cpp [hunk 1]
```diff
@@ -0,0 +1,38 @@
+// Test that -print-library-module-manifest-path finds the correct file.
+
+// REQUIRES: x86-registered-target
```
Comment (ChuanqiXu9) [inline clang/test/Driver/modules-print-library-module-manifest-path.cpp:3]:
```suggestion
// FIXME: ....
// REQUIRES: x86-registered-target
```

I feel better with a FIXME to remind us to remove this later.

**Review by ChuanqiXu9 — APPROVED** [review]
LGTM.

**Comment by ChuanqiXu9** [comment]
@mordante if we want this for 18, we need to land and backport it in this week.

**Comment by mordante** [comment]
> @mordante if we want this for 18, we need to land and backport it in this week.

Thanks for the reminder, it got a bit under my radar. I'll ask a backport tomorrow unless a CI starts to complain again.

**Comment by kaz7** [comment]
Unfortunately, this is breaking VE buildbot again.  I'll check the source of your problem, but can you revert your patch once?

https://lab.llvm.org/buildbot/#/builders/91/builds/23291

**Comment by kaz7** [comment]
Reverted.  It is possible to reproduce erros by following commands.

```
$ mkdir build
$ cd build
$ cmake -G Ninja -DCMAKE_BUILD_TYPE=Release ../llvm-project/llvm -DLLVM_ENABLE_PROJECTS=clang -DLLVM_ENABLE_RUNTIMES="libcxx;libcxxabi;libunwind"
$ ninja
$ ninja check-clang
********************
********************
Failed Tests (1):
  Clang :: Driver/modules-print-library-module-manifest-path.cpp
```

**Comment by kaz7** [comment]
The point is to build libraries and test clang.  The `TC.getFilePaths` is constructed from following items in `ToolChain::ToolChain`

1. `BUILD/bin/../lib/x86_64-unknown-linux-gnu`
2. `%t/Inputs/usr/lib/x86_64-linux-gnu`
3. `%t/Inputs/usr/lib`

Yes.  It contains 1st item at run time if you build not only clang but also libraries.  And `GetStdModuleManifestPath` find `BUILD/bin/../lib/x86_64-unknown-linux-gnu/libc++.so` at first.  However, `modules.json` is not in that directory.  It is in `%t/Inputs/usr/lib/x86_64-linux-gnu/modules.json`.

Hope this helps you.

**Comment by mordante** [comment]
Thanks for the revert and additional information. I was expecting that the test should not be affected by files outside of the `%t/Inputs` directory. I will test this locally, it might take a few days before I have time.

**Comment by mordante** [comment]
> The point is to build libraries and test clang at once. The `TC.getFilePaths` is constructed from following items in `ToolChain::ToolChain`
> 
>     1. `BUILD/bin/../lib/x86_64-unknown-linux-gnu`
> 
>     2. `%t/Inputs/usr/lib/x86_64-linux-gnu`
> 
>     3. `%t/Inputs/usr/lib`
> 
> 
> Yes. It contains 1st item at run time if you build libraries. As a result, `GetStdModuleManifestPath` find `BUILD/bin/../lib/x86_64-unknown-linux-gnu/libc++.so` first. However, `modules.json` is at `%t/Inputs/usr/lib/x86_64-linux-gnu/modules.json`.
> 
> Hope this helps you.

I can indeed reproduce this. I'm not sure whether this is the expected result. There is no real documentation for `-sysroot`, but it is intended for cross-compilation. @ChuanqiXu9 since you suggested this test approach, do you know whether this is the expected behavior of `-sysroot` ?

**Comment by ChuanqiXu9** [comment]
> @ChuanqiXu9 since you suggested this test approach

It looks like that I failed to understand VE is under X86 also...

> do you know whether this is the expected behavior of -sysroot ?

I am not sure. This is surprising to me too.



> The point is to build libraries and test clang at once. The `TC.getFilePaths` is constructed from following items in `ToolChain::ToolChain`
> 
> 1. `BUILD/bin/../lib/x86_64-unknown-linux-gnu`
> 2. `%t/Inputs/usr/lib/x86_64-linux-gnu`
> 3. `%t/Inputs/usr/lib`
> 
> Yes. It contains 1st item at run time if you build libraries. As a result, `GetStdModuleManifestPath` find `BUILD/bin/../lib/x86_64-unknown-linux-gnu/libc++.so` first. However, `modules.json` is at `%t/Inputs/usr/lib/x86_64-linux-gnu/modules.json`.
> 
> Hope this helps you.

It looks like this is not special to VE but reproducible for every constructing the clang and libc++ at once. So  I'll try to look at it if I can reproduce it locally.

**Comment by ChuanqiXu9** [comment]
I sent the new PR in https://github.com/llvm/llvm-project/pull/84881. See the comments there for details.

**Pull request merged**